src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,585 @@
+        return;
+    }
+    auto recover_task = [this, root_dir]()
+    {
+        auto callback = [this](RemoteCacheController * cntrl) { this->total_size += cntrl->size(); };
+        std::lock_guard lock(this->mutex);
+        // two level dir. /<first 3 chars of path hash code>/<path hash code>
+        recoverCachedFilesMeta(root_dir, 1, 2, callback);
+        this->inited = true;
+        LOG_TRACE(this->log, "recovered from disk ");
+    };
+    GetThreadPool()->scheduleOrThrow(recover_task);
+}
+
+std::filesystem::path RemoteReadBufferCache::calculateLocalPath(const RemoteFileMeta & meta)
+{
+    std::string full_path = meta.schema + ":" + meta.cluster + ":" + meta.path;
+    UInt128 hashcode = sipHash128(full_path.c_str(), full_path.size());
+    std::string hashcode_str = getHexUIntLowercase(hashcode);
+    return std::filesystem::path(local_path_prefix) / hashcode_str.substr(0,3) / hashcode_str;
+}
+
+std::tuple<std::shared_ptr<LocalCachedFileReader>, RemoteReadBufferCacheError> RemoteReadBufferCache::createReader(
+    const RemoteFileMeta &remote_file_meta,
+    std::shared_ptr<ReadBuffer> & readbuffer)
+{
+    // If something is wrong on startup, rollback to read from the original ReadBuffer
+    if (!hasInitialized())
+    {
+        LOG_ERROR(log, "RemoteReadBufferCache has not initialized");
+        return {nullptr, RemoteReadBufferCacheError::NOT_INIT};
+    }
+
+    auto remote_path = remote_file_meta.path;
+    const auto & file_size = remote_file_meta.file_size;
+    const auto & last_modification_timestamp = remote_file_meta.last_modification_timestamp;
+    auto local_path = calculateLocalPath(remote_file_meta);
+    std::lock_guard lock(mutex);
+    auto cache_iter = caches.find(local_path);
+    if (cache_iter != caches.end())
+    {
+        // if the file has been update on remote side, we need to redownload it
+        if (cache_iter->second.cache_controller->getLastModificationTimestamp() != last_modification_timestamp)
+        {
+            LOG_TRACE(log,
+                "remote file has been updated. " + remote_path + ":" + std::to_string(cache_iter->second.cache_controller->getLastModificationTimestamp()) + "->"
+                    + std::to_string(last_modification_timestamp));
+            cache_iter->second.cache_controller->markInvalid();
+        }
+        else
+        {
+            // move the key to the list end
+            keys.splice(keys.end(), keys, cache_iter->second.key_iterator);
+            return {std::make_shared<LocalCachedFileReader>(cache_iter->second.cache_controller.get(), file_size), RemoteReadBufferCacheError::OK};
+        }
+    }
+
+    auto clear_ret = clearLocalCache();
+    cache_iter = caches.find(local_path);
+    if (cache_iter != caches.end())
+    {
+        if (cache_iter->second.cache_controller->isValid())
+        {
+            // move the key to the list end, this case should not happen?
```

Comment:
Why should not happen? For example if this method is called for the same file path concurrently? Actually to me it now seems that there are a lot of possible races near this `isValid` and `markInvalid` checks.

Reply:
the valid flag in cache controller is only accessed in RemoteReadBufferCache in the mutex scope